### DISCUSSION TITLE
Measuring recent performance gains

### DISCUSSION THREAD

**Discussion by vkoskiv** [VKoskiv/c-ray#74] [Show and tell]
This issue is here to compile a list of measurements from the recent, rather massive, performance gains achieved with advice and code improvements from @madmann91 

I'll post more of my findings here as I do more testing in the coming days.
Ideally, these measurements should have three variants. **original**, **IPO enabled** and **IPO with the new BVH enabled**.
This should give indication as to the performance gains attained from the new BVH implementation, and isolate that from the large performance gain attained with interprocedural optimization.
All comparisons here are performed with the same exact scene and same settings. We could also perform multiple runs and average them to negate any temporary performance fluctuations of the operating system.

This is an incomplete measurement I performed yesterday on my `Core i5-6600K` desktop system running macOS:

`Original k- d tree, no IPO, input/hdr.json @2560x1600, 6000spp, 30 bounces`: 8h 40min
`Original k-d tree, IPO enabled, input/hdr.json @2560x1600, 6000spp, 30 bounces`: 7h 40min 
`New BVH, IPO enabled, input/hdr.json @2560x1600, 6000spp, 30 bounces`: 2h 15min

As you can see, massive performance gains. This measurement lacks the test with the old k-d trees + IPO enabled to truly compare k-d tree vs BVH, I'm performing that measurement at the moment.

One more note, C-ray encodes some render information into the tEXt chunks of PNG files. These can be observed with `imagemagick` as follows:
`identify -verbose /path/to/rendered_0000.png`

You should get output resembling the following:

```
...
Properties:
    C-ray Bounces: 30
    C-ray RenderTime: 2h 15m
    C-ray Samples: 6000
    C-ray Source: https://github.com/vkoskiv/c-ray
    C-ray SysInfo: x86_64 Proteus 16.7.0 Darwin Darwin Kernel Version 16.7.0: Tue Jan 30 11:27:06 PST 2018; root:xnu-3789.73.11~1/RELEASE_X86_64
    C-ray Threads: 6
    C-ray Version: C-ray v0.6.3 [8468d9e4], © 2015-2020 Valtteri Koskivuori
    date:create: 2020-06-07T22:21:18+00:00
    date:modify: 2020-06-07T21:07:31+00:00
    LodePNG: 20191109
...
```

**Comment by odil24** [comment]
Sorry for offtopic here. Didn't would like to create new issue. Please, can you add me to your telegram cuz somebody pushed me As Spam on telegram. Are you remember me?

**Comment by odil24** [comment]
![Screenshot_2020-06-09-14-36-24-1](https://user-images.githubusercontent.com/64298012/84131964-ba737d00-aa5e-11ea-9bfb-6e056c3098c5.png)

**Comment by odil24** [comment]
If you add me i can continue to discussion
